dxr/query.py

```diff
@@ -179,52 +172,70 @@ def direct_result(self):
                     return None
 
 
-query_grammar = Grammar(ur'''
-    query = _ terms
-    terms = term*
-    term = not_term / positive_term
-    not_term = not positive_term
-    positive_term = filtered_term / text
 
-    # A term with a filter name prepended:
-    filtered_term = maybe_plus filter ":" text
+def cached(c):
+    def cache_f(f):
+        def inner(enabled_plugins):
+            key = tuple(enabled_plugins)
+            if key in c:
+                return c[key]
+
+            result = f(enabled_plugins)
+            c[key] = result
+            return result
+
+        return inner
+    return cache_f
 
-    # Bare or quoted text, possibly with spaces. Not empty.
-    text = (double_quoted_text / single_quoted_text / bare_text) _
 
-    filter = ~r"''' +
-        # regexp, function, etc. No filter is a prefix of a later one. This
-        # avoids premature matches.
-        '|'.join(sorted((re.escape(filter_name) for
-                                filter_name, filters in
-                                FILTERS_NAMED.iteritems() if
-                                filters[0].description),
-                        key=len,
-                        reverse=True)) + ur'''"
+query_grammar_cache = {}
+@cached(query_grammar_cache)
```

Comment:
Much better. I did another round of simplification so you no longer have to pass in the cache from outside as well.